File: gto-support-db-stream/src/main/java/org/ccci/gto/android/common/db/StreamDao.java
package org.ccci.gto.android.common.db;

import com.annimon.stream.Stream;

import androidx.annotation.NonNull;

/**
 * Stream DAO interface relying on backwards compatible Lightweight Stream API.
 */
public interface StreamDao {
    @NonNull
    <T> Stream<T> streamCompat(@NonNull Query<T> query);

    class StreamHelper {
        @NonNull
        public static <T> Stream<T> stream(@NonNull final Dao dao, @NonNull final Query<T> query) {
            return Stream.of(dao.get(query));
        }
    }
}

File: gto-support-db-stream/src/main/java/org/ccci/gto/android/common/db/StreamDao.kt
package org.ccci.gto.android.common.db

import com.annimon.stream.Stream

/**
 * Stream DAO interface relying on backwards compatible Lightweight Stream API.
 */
interface StreamDao : Dao {
    fun <T : Any> streamCompat(query: Query<T>): Stream<T> = Stream.of(get(query))

    object StreamHelper {
        @JvmStatic
        @Deprecated("Since v3.4.0, use default implementation of streamCompat instead.")
        fun <T : Any> stream(dao: Dao, query: Query<T>): Stream<T> = Stream.of(dao.get(query))
    }
}
